cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -139,9 +148,60 @@ func (u *upgrader) init() error {
 	if err != nil {
 		return daisy.Errf("Failed to create GCE client: %v", err)
 	}
+
+	err = u.initLogger()
+	if err != nil {
+		return daisy.Errf("Failed to initialize logger: %v", err)
+	}
+
 	return nil
 }
 
+func (u *upgrader) initLogger() error {
+	if u.CloudLogsDisabled || u.ProjectPtr != nil {
+		return nil
+	}
+
+	loggingOptions := []option.ClientOption{option.WithCredentialsFile(u.Oauth)}
+	cloudLoggingClient, err := logging.NewClient(u.ctx, *u.ProjectPtr, loggingOptions...)
+	if err != nil {
+		return err
+	}
+
+	if err := cloudLoggingClient.Ping(u.ctx); err != nil {
+		fmt.Println("Failed to initialize cloud logger: ", err)
+	} else {
+		cloudLogName := fmt.Sprintf("windows-upgrade-%s", u.id)
+		u.cloudLogger = cloudLoggingClient.Logger(cloudLogName)
+		go func() {
+			for {
+				time.Sleep(5 * time.Second)
+				u.cloudLogger.Flush()
+			}
+		}()
+	}
+	return nil
+}
+
+func (u *upgrader) logf(s string, a ...interface{}) {
+	e := &daisy.LogEntry{
+		LocalTimestamp: time.Now(),
+		WorkflowName:   "windows-upgrade",
+		Message:        fmt.Sprintf(s, a...),
+	}
+	if u.Logger != nil {
+		// Use u.Logger to write to 3 destinations: stdout, CloudLogging and daisy.log(gcs).
+		u.Logger.WriteLogEntry(e)
```

Comment:
You've got a (potential) race condition here that will cause your program to crash.

Reply:
removed all log changes